p4-16/spec/P4-16-spec.mdk

```diff
@@ -6882,11 +6921,219 @@ control ingress(inout Header_t h, inout Meta_t m,
+p = prio[0];
+for (int j = 1; j < n; j += 1) {
+    if (prio_specified[j]) {
+        p = prio[j];
+    } else {
+        if (largest_priority_wins == true) {
+            p -= priority_delta;
+        } else {
+            p += priority_delta;
+        }
+        prio[j] = p;
+    }
+}
+~ End P4Pseudo
+
+This is the end of the first step: determining entry priorities.
+
+In the second step, the compiler issues errors for out of range
+priority values, and/or warnings for certain combinations of entry
+priorities that might be unintended by the developer, unless the
+developer explicitly disables those warnings.
+
+If any priority values are negative, or larger than the maximum
+supported value, that is a compile time error.
+
+If the annotation `@noWarn("duplicate_priorities")` is _not_ used on
+the `entries` table property, then the compiler issues a warning if
+any two entries for the same table have equal priority values.  Both
+P4Runtime and TDI leave it unspecified which entry is the winner if a
+lookup key matches multiple keys that all have the same priority,
+hence a warning is useful to less experienced developers that are
+unfamiliar with this unspecified behavior.
+
+If the annotation `@noWarn("duplicate_priorities")` is used on the
+`entries` table property, then no warnings of this type are ever
+issued by the compiler.  Using equal priority values for multiple
+entries in the same table is sometimes useful in reducing the number
+of hardware updates required when adding entries to such a table.
+
+If the annotation `@noWarn("entries_out_of_priority_order")` is _not_
+used on the `entries` table property, then the compiler issues a
+warning if:
+
+- If `largest_priority_wins` is `true` for the table, and there is any
+  pair of consecutive entries where `prio[j] < prio[j+1]`, then a
+  warning is issued for that pair of entries.
+- If `largest_priority_wins` is `false` for the table, and there is any
+  pair of consecutive entries where `prio[j] > prio[j+1]`, then a
+  warning is issued for that pair of entries.
+
+This warning is useful to developers that want the order that entries
+appear in the source code to match the relative priority of entries in
+the target device.
+
+If the annotation `@noWarn("entries_out_of_priority_order")` is used
+on the `entries` table property, then no warnings of this type are
+ever issued by the compiler for this table.  This option is provided
+for developers who explicitly choose to specify entries in an order
+that does not match their relative priority order.
+
+The following example is the same as the first example in section
+[#sec-entries], except for the definition of table `t_exact_ternary`
+shown below.
+
+~ Begin P4Example
+table t_exact_ternary {
+    key = {
+        h.h.e : exact;
+        h.h.t : ternary;
+    }
+
+    actions = {
+        a;
+        a_params;
+    }
+
+    default_action = a;
+
+    largest_priority_wins = false;
+    priority_delta = 10;
```

Comment:
We do have a `@priority` annotation already implemented in p4c, yes.  The reason for the suggested syntax NOT using that here is the same general "prefer NOT to use annotations for things that can change the semantics of the program if they were removed" guidance that we often hear in the LDWG meetings when annotations are proposed.

Reply:
There is a `@priority` annotation implemented in p4c since a 2017 commit to p4c by Calin Cascaval, but it is described nowhere in the language spec, and I believe never has.

Do you want me to deprecate something not mentioned in the spec, in the spec?

Reply:
Maybe you can find out if the barefoot backend uses them. We care about implementations even if they are not conformant with the spec.

Reply:
Checking with @slicking @sayanb @vgurevich to see if they know.

Reply:
I checked with @slicking and the only uses he found of `@priority` in our code were test programs copied from public p4c.

Reply:
In fact, this is what the compiler outputs if you use `@priority`:
```
table-entries-priority-bmv2.p4(79): [--Wwarn=deprecated] warning: The @priority annotation on Entry is not part of the P4 specification, nor of the P4Runtime specification, and will be ignored
            0x1181 &&& 0xF00F : a_with_control_params(3) @priority(1);
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

So I guess they are already deprecated.